projects/plutus/src/ltc_afe.c

```diff
@@ -0,0 +1,310 @@
+  LTC6804_RDCOMM_RESERVED
+};
+
+static uint8_t s_voltage_reg[NUM_LTC_AFE_VOLTAGE_REGISTER] = {
+  LTC_AFE_REGISTER_CELL_VOLTAGE_A,
+  LTC_AFE_REGISTER_CELL_VOLTAGE_B,
+  LTC_AFE_REGISTER_CELL_VOLTAGE_C,
+  LTC_AFE_REGISTER_CELL_VOLTAGE_D
+};
+
+static void prv_wakeup_idle(const LTCAFESettings *afe) {
+  gpio_set_pin_state(&afe->cs, GPIO_STATE_LOW);
+  delay_us(2);
+  gpio_set_pin_state(&afe->cs, GPIO_STATE_HIGH);
+}
+
+static StatusCode prv_read_register(const LTCAFESettings *afe,
+                                    LTCAFERegister reg, uint8_t *data, size_t len) {
+  if (reg > NUM_LTC_AFE_REGISTER) {
+    return status_code(STATUS_CODE_INVALID_ARGS);
+  }
+
+  uint16_t reg_cmd = s_read_reg_cmd[reg];
+  uint8_t cmd[4] = { 0 };
+
+  cmd[0] = (uint8_t)(reg_cmd >> 8);
+  cmd[1] = (uint8_t)(reg_cmd & 0xFF);
+
+  uint16_t cmd_pec = crc15_calculate(cmd, 2);
+  cmd[2] = (uint8_t)(cmd_pec >> 8);
+  cmd[3] = (uint8_t)(cmd_pec & 0xFF);
+
+  return spi_exchange(afe->spi_port, cmd, 4, data, len);
+}
+
+static StatusCode prv_read_voltage(LTCAFESettings *afe, LTCAFEVoltageRegister reg, uint8_t *data) {
+  if (reg > NUM_LTC_AFE_VOLTAGE_REGISTER) {
+    return status_code(STATUS_CODE_INVALID_ARGS);
+  }
+
+  // 6 bytes in register + 2 bytes for PEC
+  size_t len = ((LTC6804_CELLS_IN_REG * 2) + 2) * PLUTUS_AFE_DEVICES_IN_CHAIN;
+
+  return prv_read_register(afe, s_voltage_reg[reg], data, len);
+}
+
+// start cell voltage conversion
+static void prv_trigger_adc_conversion(const LTCAFESettings *afe) {
+  uint8_t mode = (uint8_t)((afe->adc_mode + 1) % 3);
+  // ADCV command
+  uint16_t adcv = LTC6804_ADCV_RESERVED | LTC6804_ADCV_DISCHARGE_NOT_PERMITTED
+                  | LTC6804_CNVT_CELL_ALL | (mode << 7);
+  uint8_t cmd[4] = { 0 };
+
+  cmd[0] = (uint8_t)(adcv >> 8);
+  cmd[1] = (uint8_t)(adcv & 0xFF);
+
+  uint16_t cmd_pec = crc15_calculate(cmd, 2);
+  cmd[2] = (uint8_t)(cmd_pec >> 8);
+  cmd[3] = (uint8_t)(cmd_pec);
+
+  prv_wakeup_idle(afe);
+
+  spi_exchange(afe->spi_port, cmd, 4, NULL, 0);
+
+  // wait for conversions to finish
+  delay_ms(100);
+}
+
+static void prv_trigger_aux_adc_conversion(const LTCAFESettings *afe) {
+  uint8_t mode = (uint8_t)((afe->adc_mode + 1) % 3);
+  // ADAX
+  uint16_t adax = LTC6804_ADAX_RESERVED | LTC6804_ADAX_GPIO1 | (mode << 7);
+  uint8_t cmd[4] = { 0 };
+  cmd[0] = (uint8_t)(adax >> 8);
+  cmd[1] = (uint8_t)(adax & 0xFF);
+
+  uint16_t cmd_pec = crc15_calculate(cmd, 2);
+  cmd[2] = (uint8_t)(cmd_pec >> 8);
+  cmd[3] = (uint8_t)(cmd_pec & 0xFF);
+
+  spi_exchange(afe->spi_port, cmd, 4, NULL, 0);
+
+  // wait for conversions to finish
+  delay_ms(10);
+}
+
+// write config to all devices
+static StatusCode prv_write_config(const LTCAFESettings *afe, uint8_t gpio_enable_pins) {
+  // see p.54 in datasheet
+  // (2 bits for WRCFG + 2 bits for WRCFG PEC) +
+  // (6 bits for CFGR + 2 bits for CFGR PEC) * PLUTUS_AFE_DEVICES_IN_CHAIN
+  uint8_t configuration_cmd[(2 + 2) + (6 + 2) * PLUTUS_AFE_DEVICES_IN_CHAIN] = { 0 };
```

Comment:
Recommend using structs instead of raw byte arrays to make it easier to understand what's going on and remove the need for all the index-related math in this function.